refactor(ivy): simplify bind instruction to reuse bindingUpdated logic

JoostK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

The pipe instruction in Ivy duplicated the logic for updating a value binding.

## What is the new behavior?

The pipe instruction now reuses common logic to update a binding.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I was looking into where a `NO_CHANGE` input value would actually come from, given that no testcases seemed to cover that situation. After looking into the compiler it became apparent that it may generate a "double binding" in the form of `bind(interpolate#(value))`, which in runtime would cause the `NO_CHANGE` to be passed into `bind`.

For now I decided to add a compiler testcase that shows that the current behavior actually generates a double binding, and a runtime testcase that verifies that this situation is handled correctly. We may choose to identify this case in the compiler and not emit the `bind` instruction in the first place, given that it takes an extra data slot for tracking the binding value and also duplicates the difference checking of the binding.

JoostK committed:
refactor(ivy): simplify bind instruction to reuse bindingUpdated logic

Added runtime and compiler testcases for interpolated bindings, which verify
that NO_CHANGE is properly handled in `bind`.

marclaval commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -35,6 +35,62 @@ describe('instructions', () => {
     elementEnd();
   }
 
+  describe('bind', () => {
```
We already have similar tests in the *integration_spec.ts* file:
- the first two exist in the `element bindings` > `elementAttribute` section, adding the check on ngDevMode there would be nice
- the third one could be added to the `text bindings` section

JoostK replied:
Hmm, to an outsider like myself I expect these kinds of tests in `instructions_spec`. Also, the `ngDevMode` assertions are typically found here instead of in the integration suite.

marclaval commented on packages/compiler/test/render3/r3_view_compiler_spec.ts:
```diff
@@ -120,5 +120,43 @@ describe('r3_view_compiler', () => {
       expectEmit(result.source, bV_call, 'Incorrect bV call');
     });
+
+    it('should generate interpolated bindings', () => {
```
Not sure about this test, I'll let @vicb or @alxhub comment on this

mhevery replied:
Yes @kara is right, this is a compile bug. Its one or the other.

marclaval commented:
The refactor looks good to me. I've left some comments about the tests

JoostK commented:
@marclaval Actually, I came to realize that this change introduces a bug, where the `bindingIndex` is not incremented for a `NO_CHANGE` value. This means that subsequent bindings will get misplaced. Given that the unit tests pass indicates that this scenario is not covered at the moment.

kara reviewed (commented):
The refactor looks good to me, but a few nits since you actually caught a compiler bug :-)

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2112,20 +2112,7 @@ function initBindings() {
-  }
-  currentView.bindingIndex++;
-  return changed ? value : NO_CHANGE;
+  return value !== NO_CHANGE && bindingUpdated(value) ? value : NO_CHANGE;
```
👍

JoostK replied:
@kara Yes! At first I almost committed the signature change to remove `NO_CHANGE` as that was my initial understanding too, but then found the compiler generating the double binding so I reverted that change :)

JoostK commented:
Let me try to actually fix the compiler, that seems like a fun task 😄

Thanks everyone for providing input!

JoostK committed:
fixup! refactor(ivy): simplify bind instruction to reuse bindingUpdated logic

It is actually illegal for bind to be passed `NO_CHANGE`, so the signature
has been changed to reflect that. `bindingUpdated` correctly asserts for
not receiving `NO_CHANGE`.

JoostK committed:
fixup! refactor(ivy): simplify bind instruction to reuse bindingUpdated logic

It is actually illegal for bind to be passed `NO_CHANGE`, so the signature
has been changed to reflect that. `bindingUpdated` correctly asserts for
not receiving `NO_CHANGE`.

JoostK commented:
CircleCI doesn't seem kind to me today. Anyway, `bind` does not accept `NO_CHANGE` any longer.

alexzuza commented on packages/core/test/render3/render_util.ts:
```diff
@@ -57,9 +57,9 @@ export class TemplateFixture extends BaseFixture {
-   *          `if (creationMode) { ... } __here__`.
+   *          `if (rf & RenderFlags.Create) { __here__ }`.
+   * @param updateBlock Optional instructions which go into the update block:
+   *          `if (rf & RenderFlags.Update) { __here__ }`.
```
@kara @mhevery Seems there is another compiler bug related to HostBinding implementation:

Compiler generates code for load data (`ɵld`)
https://github.com/angular/angular/blob/b7c417f618f1aa27fe1613129711d1380ce71dce/packages/compiler/test/render3/r3_compiler_compliance_spec.ts#L294

while core package uses `loadDirective`(`ɵd`)
https://github.com/angular/angular/blob/b7c417f618f1aa27fe1613129711d1380ce71dce/packages/core/test/render3/compiler_canonical/component_directives_spec.ts#L103

As a result Hostbinding won't work propertly since it tries to find property on LNode but should look at directive instance

kara replied:
We used to generate `ld` in a previous implementation, so it seems like the compiler hasn't caught up yet. Can you create a separate Issue for this? Seems outside the scope of this PR.

kara reviewed (approved):
LGTM

mhevery commented:
http://test/OCL:198582492:BASE:198582512:1527701873636:d22f3dcb

JoostK commented:
@mhevery I like what you're doing with the timezone labels. Thanks for the rebase btw 👍

vicb commented:
Note to self: Rebase once master is green.
